### PR TITLE
fix(source): init split state with persisted state and fix cdc split update

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fix some issues for Source
1. Previously source executor doesn't initialize the in-memory split state with persisted state, this PR fix it
2. Prevent `snapshot_done` of cdc split being reset unexpectedly by only updating when it is `false`
3. replace `copy_with_offset` with `update_with_offset` for source split

Tested cdc source on local env.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/connector/src/source/cdc/split.rs [hunk 1]
```diff
@@ -92,22 +92,18 @@ impl MySqlCdcSplit {
...
-            split_id: self.inner.split_id,
-            start_offset: Some(start_offset),
-            snapshot_done,
-        };
-        Self { inner: split }
+    pub fn update_with_offset(&mut self, start_offset: String) {
+        let mut snapshot_done = self.inner.snapshot_done;
+        if !snapshot_done {
+            let dbz_offset: DebeziumOffset = serde_json::from_str(&start_offset)
+                .unwrap_or_else(|e| panic!("invalid cdc offset: {}, error: {}", start_offset, e));
```
Comment (tabVersion) [inline src/connector/src/source/cdc/split.rs:99]: also print split_id here?

File: src/connector/src/source/cdc/split.rs [hunk 2]
```diff
@@ -124,27 +120,19 @@ impl PostgresCdcSplit {
...
-
-        let server_addr = self.server_addr.clone();
-        Self {
-            inner: split,
-            server_addr,
+    pub fn update_with_offset(&mut self, start_offset: String) {
+        let mut snapshot_done = self.inner.snapshot_done;
+        if !snapshot_done {
+            let dbz_offset: DebeziumOffset = serde_json::from_str(&start_offset)
+                .unwrap_or_else(|e| panic!("invalid cdc offset: {}, error: {}", start_offset, e));
```
Comment (tabVersion) [inline src/connector/src/source/cdc/split.rs:127]: a little -1 for panic here, shall we throw an error instead?

**Comment by tabVersion** [comment]
> Previously source executor doesn't initialize the in-memory split state with persisted state, this PR fix it

https://github.com/risingwavelabs/risingwave/blob/a9fe6638d1f8057f290b194a41ad05b2fec53dc6/src/stream/src/executor/source/source_executor.rs#L231-L239 
this part loads persisted states from state table
Rest LGTM

**Review by tabVersion — APPROVED** [review]

File: src/connector/src/source/pulsar/split.rs [hunk 3]
```diff
@@ -38,6 +38,10 @@ impl PulsarSplit {
             start_offset,
         }
     }
+
+    pub fn update_with_offset(&mut self, _start_offset: String) {
+        unimplemented!("pulsar split")
+    }
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/connector/src/source/pulsar/split.rs:44]: Seems the latest code doesn't push. I‘ll fit it.

File: src/rpc_client/src/connector_client.rs [hunk 4]
```diff
@@ -197,11 +197,6 @@ impl ConnectorClient {
         properties: HashMap<String, String>,
         snapshot_done: bool,
     ) -> Result<Streaming<GetEventStreamResponse>> {
-        tracing::info!(
-            "start cdc source properties: {:?}, snapshot_done: {}",
-            properties,
-            snapshot_done
-        );
```
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/rpc_client/src/connector_client.rs:204]: It will print sensitive info in the properties.

**Comment by StrikeW** [comment]
> > Previously source executor doesn't initialize the in-memory split state with persisted state, this PR fix it
> 
> https://github.com/risingwavelabs/risingwave/blob/a9fe6638d1f8057f290b194a41ad05b2fec53dc6/src/stream/src/executor/source/source_executor.rs#L231-L239
> 
> 
> this part loads persisted states from state table
> Rest LGTM

You can see that the `core.stream_source_splits` is initialized only with the split state from the first barrier.
https://github.com/risingwavelabs/risingwave/blob/def44629ea280110f97ebbde9e7cdb1d4b22d280/src/stream/src/executor/source/source_executor.rs#L364-L386

File: src/connector/src/source/cdc/split.rs (hunk 2 above)
Comment (StrikeW) [reply src/connector/src/source/cdc/split.rs]: fixed

**Pull request merged**